### PR TITLE
Support filter and nested field in faiss engine radial search

### PR DESCRIPTION
**Pull request by junqiu-lei**
### Description
This PR will allow user to use filter and nested field in faiss engine radial search by reusing existing logics from top-k search. 
* The custom patch for range_search is reused from [faiss search patch](https://github.com/opensearch-project/k-NN/blame/main/jni/patches/faiss/0001-Custom-patch-to-support-multi-vector.patch#L136-L158). Please check [this](https://github.com/junqiu-lei/faiss/commit/b36a28778a493df46a7f40b03cc5b426e94a4ba1) commit for the patch updated code in faiss.
* In jni faiss_wrapper.cpp, the params set up process is reused from existing search function.
* IT tests added for both nested field and filter queries.

### Issues Resolved
Part of #814  
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jmazanec15** [comment]
@junqiu-lei For this patch, did we create a PR in faiss to upstream it?

**Comment by junqiu-lei** [comment]
> @junqiu-lei For this patch, did we create a PR in faiss to upstream it?

@jmazanec15 No, I mainly just followed the existing patch for search function https://github.com/opensearch-project/k-NN/blame/main/jni/patches/faiss/0001-Custom-patch-to-support-multi-vector.patch#L136-L158

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased 2.x](https://github.com/opensearch-project/k-NN/compare/2.13...2.x)
 ### Features
-* Support radial search in k-NN plugin [#1617](https://github.com/opensearch-project/k-NN/pull/1617)
+* Support radial search in k-NN plugin [#814](https://github.com/opensearch-project/k-NN/issues/814)
```
Comment (martin-gaievski) [inline CHANGELOG.md:17]: why we're changing PR ref?

File: src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java [hunk 2]
```diff
@@ -108,33 +110,33 @@ public KNNQueryBuilder k(Integer k) {
...
      *
-     * @param max_distance the max_distance threshold for the nearest neighbours
+     * @param maxDistance the maxDistance threshold for the nearest neighbours
      */
-    public KNNQueryBuilder maxDistance(Float max_distance) {
-        if (max_distance == null) {
-            throw new IllegalArgumentException("[" + NAME + "] requires max_distance to be set");
+    public KNNQueryBuilder maxDistance(Float maxDistance) {
+        if (maxDistance == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires maxDistance to be set");
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java:119]: Why not `String.format`? Same comment for multiple similar lines where you combine exception messages

File: CHANGELOG.md (hunk 1 above)
Comment (junqiu-lei) [reply CHANGELOG.md:17]: I intended to use the feature issue as entrypoint here, will update to have a new line for this PR.

File: src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java (hunk 2 above)
Comment (junqiu-lei) [reply src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java:119]: Thanks will update

**Comment by jmazanec15** [comment]
@junqiu-lei one minor comment but other than that it looks good

File: jni/patches/faiss/0003-Custom-patch-to-support-range-search-params.patch
Comment (junqiu-lei) [inline jni/patches/faiss/0003-Custom-patch-to-support-range-search-params.patch:1]: Please check [this](https://github.com/junqiu-lei/faiss/commit/b36a28778a493df46a7f40b03cc5b426e94a4ba1) commit for the patch diff.

File: CHANGELOG.md (hunk 1 above)
Comment (junqiu-lei) [reply CHANGELOG.md]: Updated

File: jni/src/faiss_wrapper.cpp [hunk 3]
```diff
@@ -27,6 +27,7 @@
 #include <jni.h>
 #include <string>
 #include <vector>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (junqiu-lei) [reply jni/src/faiss_wrapper.cpp]: Updated

**Review by jmazanec15 — APPROVED** [review]

**Review by martin-gaievski — APPROVED** [review]
Looks good to me, thanks

**Pull request merged**